File: proxy/douyin-proxy.js
// DouYin live start time proxy
// Deploy alongside biliLive-tools on EC2.
// CF Worker calls this to get real create_time from DouYin mobile API.
//
// Usage: node douyin-proxy.js
// Listens on port 3002 by default (configurable via PORT env var)

import http from "http";

const PORT = process.env.PORT || 3002;
const PASSKEY = process.env.PROXY_PASSKEY || "";

async function fetchDouyinCreateTime(secUserId) {
  const params = new URLSearchParams({
    app_id: "1128",
    live_id: "1",
    verifyFp: "",
    room_id: "2",
    type_id: "0",
    sec_user_id: secUserId,
  });
  const url = "https://webcast.amemv.com/webcast/room/reflow/info/?" + params.toString();
  const res = await fetch(url, {
    headers: { "User-Agent": "Mozilla/5.0 (Linux; Android 12) AppleWebKit/537.36" },
  });
  if (!res.ok) return { ok: false, error: "upstream " + res.status };
  const data = await res.json();
  if (data.status_code !== 0) return { ok: false, error: "douyin " + data.status_code };
  const room = data?.data?.room;
  if (!room || room.status !== 2) return { ok: true, living: false, create_time: 0 };
  return { ok: true, living: true, create_time: room.create_time || 0 };
}

const server = http.createServer(async (req, res) => {
  // Health check
  if (req.url === "/health") {
    res.writeHead(200, { "Content-Type": "application/json" });
    res.end('{"ok":true}');
    return;
  }

  // Auth check
  if (PASSKEY && req.headers.authorization !== PASSKEY) {
    res.writeHead(401, { "Content-Type": "application/json" });
    res.end('{"ok":false,"error":"unauthorized"}');
    return;
  }

  // GET /create-time?uid=xxx
  if (req.method === "GET" && req.url.startsWith("/create-time")) {
    const urlObj = new URL(req.url, "http://localhost");
    const uid = urlObj.searchParams.get("uid") || "";
    if (!uid) {
      res.writeHead(400, { "Content-Type": "application/json" });
      res.end('{"ok":false,"error":"missing uid"}');
      return;
    }
    try {
      const result = await fetchDouyinCreateTime(uid);
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify(result));
    } catch (e) {
      res.writeHead(502, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ ok: false, error: String(e) }));
    }
    return;
  }

  // GET /create-times — batch: POST body = {"uids":["uid1","uid2"]}
  if (req.method === "POST" && req.url === "/create-times") {
    let body = "";
    for await (const chunk of req) body += chunk;
    try {
      const { uids } = JSON.parse(body);
      const results = await Promise.all(
        (uids || []).map(async (uid) => {
          try {
            const r = await fetchDouyinCreateTime(uid);
            return { uid, ...r };
          } catch (e) {
            return { uid, ok: false, error: String(e) };
          }
        })
      );
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ ok: true, results }));
    } catch (e) {
      res.writeHead(400, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ ok: false, error: String(e) }));
    }
    return;
  }

  res.writeHead(404, { "Content-Type": "application/json" });
  res.end('{"ok":false,"error":"not found"}');
});

server.listen(PORT, () => {
  console.log("DouYin proxy listening on port " + PORT);
});

File: proxy/douyin-proxy.py
#!/usr/bin/env python3
"""DouYin live start time proxy.
Deploy on Huawei Cloud (domestic IP) to bypass DouYin's cloud IP blocking.

Usage: python3 douyin-proxy.py
Env vars: PORT (default 3002), PROXY_PASSKEY (optional auth)
"""

import os, json, urllib.request, urllib.parse
from http.server import HTTPServer, BaseHTTPRequestHandler

PORT = int(os.environ.get("PORT", 3002))
PASSKEY = os.environ.get("PROXY_PASSKEY", "")

DOUYIN_UA = "Mozilla/5.0 (Linux; Android 12) AppleWebKit/537.36"

def fetch_create_time(sec_user_id):
    params = urllib.parse.urlencode({
        "app_id": "1128", "live_id": "1", "verifyFp": "",
        "room_id": "2", "type_id": "0", "sec_user_id": sec_user_id,
    })
    url = f"https://webcast.amemv.com/webcast/room/reflow/info/?{params}"
    req = urllib.request.Request(url, headers={"User-Agent": DOUYIN_UA})
    with urllib.request.urlopen(req, timeout=10) as resp:
        data = json.loads(resp.read())
    if data.get("status_code") != 0:
        return {"ok": False, "error": f"douyin {data.get('status_code')}"}
    room = (data.get("data") or {}).get("room")
    if not room or room.get("status") != 2:
        return {"ok": True, "living": False, "create_time": 0}
    return {"ok": True, "living": True, "create_time": room.get("create_time", 0)}

class Handler(BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path == "/health":
            self._respond(200, {"ok": True})
            return
        if PASSKEY and self.headers.get("Authorization") != PASSKEY:
            self._respond(401, {"ok": False, "error": "unauthorized"})
            return
        if self.path.startswith("/create-time"):
            uid = urllib.parse.parse_qs(urllib.parse.urlparse(self.path).query).get("uid", [""])[0]
            if not uid:
                self._respond(400, {"ok": False, "error": "missing uid"})
                return
            try:
                self._respond(200, fetch_create_time(uid))
            except Exception as e:
                self._respond(502, {"ok": False, "error": str(e)})
            return
        self._respond(404, {"ok": False, "error": "not found"})

    def do_POST(self):
        if PASSKEY and self.headers.get("Authorization") != PASSKEY:
            self._respond(401, {"ok": False, "error": "unauthorized"})
            return
        if self.path == "/create-times":
            body = json.loads(self.r.read(int(self.headers.get("Content-Length", 0))))
            uids = body.get("uids", [])
            results = []
            for uid in uids:
                try:
                    r = fetch_create_time(uid)
                    r["uid"] = uid
                    results.append(r)
                except Exception as e:
                    results.append({"uid": uid, "ok": False, "error": str(e)})
            self._respond(200, {"ok": True, "results": results})
            return
        self._respond(404, {"ok": False, "error": "not found"})

    def _respond(self, code, obj):
        body = json.dumps(obj).encode()
        self.send_response(code)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def log_message(self, format, *args):
        pass  # suppress access logs

if __name__ == "__main__":
    server = HTTPServer(("0.0.0.0", PORT), Handler)
    print(f"DouYin proxy listening on port {PORT}")
    server.serve_forever()
